File: sdk/js/src/interfaces/IOllamaMethods.ts
import type {
  OllamaPullModelRequest,
  OllamaDeleteModelRequest,
  OllamaGenerateEmbeddingsRequest,
  OllamaGenerateCompletionRequest,
  OllamaGenerateChatCompletionRequest,
} from '../models/ollama';
import type {
  SharpAIPullModelResponse,
  OllamaLocalModel,
  OllamaGenerateEmbeddingsResult,
  OllamaGenerateCompletionResult,
  OllamaGenerateChatCompletionChunk,
  OllamaStreamingCompletionResult,
} from '../models/ollama';

/**
 * Interface for Ollama API methods.
 */
export interface IOllamaMethods {
  /**
   * Pull a model from the registry with streaming progress updates.
   * @param request - Pull model request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns AsyncGenerator yielding pull model progress updates.
   */
  pullModel(
    request: OllamaPullModelRequest,
    signal?: AbortSignal
  ): AsyncGenerator<SharpAIPullModelResponse, void, unknown>;

  /**
   * Delete a model.
   * @param request - Delete model request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns Delete model result or null on failure.
   */
  deleteModel(
    request: OllamaDeleteModelRequest,
    signal?: AbortSignal
  ): Promise<object | null>;

  /**
   * List local models.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns List of local models or null on failure.
   */
  listLocalModels(signal?: AbortSignal): Promise<OllamaLocalModel[] | null>;

  /**
   * Generate embeddings for the given input text(s).
   * @param request - The embeddings request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns Embeddings result or null on failure.
   */
  generateEmbeddings(
    request: OllamaGenerateEmbeddingsRequest,
    signal?: AbortSignal
  ): Promise<OllamaGenerateEmbeddingsResult | null>;

  /**
   * Generate text completion.
   * @param request - Generate completion request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns Generated completion result or null on failure.
   */
  generateCompletion(
    request: OllamaGenerateCompletionRequest,
    signal?: AbortSignal
  ): Promise<OllamaGenerateCompletionResult | null>;

  /**
   * Generate chat completion.
   * @param request - Generate chat completion request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns Generated chat completion result or null on failure.
   */
  generateChatCompletion(
    request: OllamaGenerateChatCompletionRequest,
    signal?: AbortSignal
  ): Promise<OllamaGenerateCompletionResult | null>;

  /**
   * Generate chat completion with streaming.
   * @param request - Generate chat completion request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns AsyncGenerator yielding streaming chat completion results.
   */
  generateChatCompletionStream(
    request: OllamaGenerateChatCompletionRequest,
    signal?: AbortSignal
  ): AsyncGenerator<OllamaGenerateChatCompletionChunk, void, unknown>;

  /**
   * Generate text completion with streaming.
   * @param request - Generate completion request.
   * @param signal - Optional AbortSignal for cancellation.
   * @returns AsyncGenerator yielding streaming completion results.
   */
  generateCompletionStream(
    request: OllamaGenerateCompletionRequest,
    signal?: AbortSignal
  ): AsyncGenerator<OllamaStreamingCompletionResult, void, unknown>;

  /** Import a GGUF model already present on the server's local filesystem (no download, no token). */
  importModel(
    request: { path: string; name?: string },
    signal?: AbortSignal
  ): Promise<object | null>;

  /** List Modelfile-equivalent presets. */
  listPresets(signal?: AbortSignal): Promise<object | null>;

  /** Create a Modelfile-equivalent preset. */
  createPreset(
    request: {
      name: string;
      model: string;
      system?: string;
      temperature?: number;
      max_tokens?: number;
      top_p?: number;
      template?: string;
      stop?: string[];
    },
    signal?: AbortSignal
  ): Promise<object | null>;

  /** Read a preset by name. */
  getPreset(name: string, signal?: AbortSignal): Promise<object | null>;

  /** Delete a preset by name. */
  deletePreset(name: string, signal?: AbortSignal): Promise<object | null>;
}

File: sdk/csharp/src/SharpAI.Sdk/Interfaces/IOllamaMethods.cs
namespace SharpAI.Sdk.Interfaces
{
    using SharpAI.Models.Ollama;
    using SharpAI.Sdk.Models;

    /// <summary>
    /// Interface for Ollama API methods.
    /// </summary>
    public interface IOllamaMethods
    {
        /// <summary>
        /// Pull a model from the registry with streaming progress updates.
        /// </summary>
        /// <param name="request">Pull model request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Async enumerable of pull model progress updates.</returns>
        IAsyncEnumerable<SharpAIPullModelResponse> PullModel(OllamaPullModelRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Delete a model.
        /// </summary>
        /// <param name="request">Delete model request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Delete model result.</returns>
        Task<object?> DeleteModel(OllamaDeleteModelRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// List local models.
        /// </summary>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>List of local models.</returns>
        Task<List<OllamaLocalModel>?> ListLocalModels(CancellationToken cancellationToken = default);

        /// <summary>
        /// Generate embeddings for the given input text(s).
        /// </summary>
        /// <param name="request">The embeddings request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Embeddings result.</returns>
        Task<OllamaGenerateEmbeddingsResult?> GenerateEmbeddings(OllamaGenerateEmbeddingsRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Generate embeddings for multiple input texts.
        /// </summary>
        /// <param name="request">The embeddings request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Embeddings result.</returns>
        Task<OllamaGenerateEmbeddingsResult?> GenerateMultipleEmbeddings(OllamaGenerateEmbeddingsRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Generate text completion.
        /// </summary>
        /// <param name="request">Generate completion request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Generated completion result.</returns>
        Task<OllamaGenerateCompletionResult?> GenerateCompletion(OllamaGenerateCompletionRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Generate chat completion.
        /// </summary>
        /// <param name="request">Generate chat completion request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Generated chat completion result.</returns>
        Task<OllamaGenerateCompletionResult?> GenerateChatCompletion(OllamaGenerateChatCompletionRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Generate chat completion with streaming.
        /// </summary>
        /// <param name="request">Generate chat completion request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Async enumerable of streaming chat completion results.</returns>
        IAsyncEnumerable<OllamaGenerateChatCompletionChunk> GenerateChatCompletionStream(OllamaGenerateChatCompletionRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Generate text completion with streaming.
        /// </summary>
        /// <param name="request">Generate completion request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Async enumerable of streaming completion results.</returns>
        IAsyncEnumerable<OllamaStreamingCompletionResult> GenerateCompletionStream(OllamaGenerateCompletionRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Import a GGUF model already present on the server's local filesystem (no download, no token).
        /// </summary>
        /// <param name="request">Import request (local path + optional name).</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The registered model details.</returns>
        Task<object?> ImportModel(SharpAIImportModelRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// List Modelfile-equivalent presets.
        /// </summary>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The preset enumeration result.</returns>
        Task<object?> ListPresets(CancellationToken cancellationToken = default);

        /// <summary>
        /// Create a Modelfile-equivalent preset.
        /// </summary>
        /// <param name="request">Preset definition.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The created preset.</returns>
        Task<object?> CreatePreset(SharpAICreatePresetRequest request, CancellationToken cancellationToken = default);

        /// <summary>
        /// Read a preset by name.
        /// </summary>
        /// <param name="name">Preset name.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The preset, or default when not found.</returns>
        Task<object?> GetPreset(string name, CancellationToken cancellationToken = default);

        /// <summary>
        /// Delete a preset by name.
        /// </summary>
        /// <param name="name">Preset name.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The delete result.</returns>
        Task<object?> DeletePreset(string name, CancellationToken cancellationToken = default);
    }
}
